doc/reference/trd-uart.md

```diff
@@ -0,0 +1,523 @@
+}
+```
+
+Methods in `Configure` can return the following error conditions:
+  - `OFF`: The underlying hardware is currently not available, perhaps
+    because it has not been initialized or in the case of a shared
+    hardware USART controller because it is set up for SPI.
+  - `INVAL`: Baud rate was set to 0.
+  - `ENOSUPPORT`: The underlying UART cannot satisfy this configuration.
+  - `FAIL`: Other failure condition.
+
+
+The UART may be unable to set the precise baud rate specified. For
+example, the UART may be driven off a fixed clock with integer
+prescalar. A call to `configure` MUST set the baud rate to the closest
+possible value to the `baud_rate` field of the `params` argument and a
+call to `set_baud_rate` MUST set the baud rate to the closest possible
+value to the `rate` argument. The `Ok` result of `set_baud_rate`
+includes the actual rate set, while an `Err(INVAL)` result means the
+requested rate is well outside the operating speed of the UART (e.g., 4MHz).
+
+
+3 `Transmit` and `TransmitClient`
+===============================
+
+The `Transmit` and `TransmitClient` traits allow a client to transmit
+bytes over the UART.
+
+```rust
+enum AbortResult {
+    Failure,
+    Success,
+}
+
+pub trait Transmit<'a> {
+    fn set_transmit_client(&self, client: &'a dyn TransmitClient);
+
+    fn transmit_buffer(
+        &self,
+        tx_buffer: &'static mut [u8],
+        tx_len: usize,
+    ) -> Result<(), (ErrorCode, &'static mut [u8])>;
+
+    fn transmit_word(&self, word: u32) -> Result<(), ErrorCode>;
+    fn transmit_abort(&self) -> Result<AbortResult, ()>;
+}
+
+pub trait TransmitClient {
+    fn transmitted_word(&self, _rval: Result<(), ErrorCode>) {}
+    fn transmitted_buffer(
+        &self,
+        tx_buffer: &'static mut [u8],
+        tx_len: usize,
+        rval: Result<(), ErrorCode>,
+    );
+}
+```
+
+The `Transmit` trait has two data paths: `transmit_word` and
+`transmit_buffer`.  The `transmit_word` method is used in narrow use
+cases where the cost and complexity of buffer management is not
+needed. Generally, software should use the `transmit_buffer`
+method. Most software implementations use DMA, such that a call to
+`transmit_buffer` triggers a single interrupt when the transfer
+completes; this saves energy and CPU cycles over per-byte transfers
+and also improves transfer speeds because hardware can keep the UART
+busy.
+
+Each byte transmitted is a data word for the UART. If the UART is
+using 8-bit data words, each data word is a byte. If the UART is using
+smaller data words, it MUST ignore the high order bits of the data
+values. For example, if the UART is using 6-bit data words and is told
+to transmit `0xff`, it will transmit `0x3f`, ignoring the first two
+bits. If a client needs to transmit data words larger than 8 bits, it
+should use `transmit_word`, as `transmit_buffer` is a buffer of 8-bit
+bytes and cannot store 9-bit values.
```

Comment:
Is this case even possible? The `Width` enum above only goes to 8 bits. I suspect 9-bit UART to be such a rarity in modern hardware perhaps that we should omit it at this stage, and then re-evaluate later what the best approach is (i.e. one would still hope to use DMA; perhaps the engine expects 16-bit halfword streams of 9-bit data? I think the 9-bit interface is a question best answered by hardware that uses it)

Reply:
https://community.silabs.com/s/article/uart-using-9-bit-mode-of-the-uart?language=en_US

There can be all kinds of weird hardware you want to connect to: I'd hate for someone who has to talk 9N1 (on a platform that can) having to start hacking peripheral drivers to do so... The fact that it's only word-level means that there isn't great support, but at least there's *something*.

`Width` should probably include 9?